rfcs/0031-bazel.md

```diff
@@ -0,0 +1,358 @@
+
+- Default local mono-builds ("native" builds) succeed on all supported
+  platforms; i.e. running build without passing any options by command
+  line or config file. Default settings just work, but may not be
+  optimal in every case.
+- Build refinements are easy and just work.  It must be easy for the
+  developer to express desired properties, and the ensuing build
+  configuration must succeed.  For example, optimization or debugging
+  flags.
+- Supported cross-platform builds must be easily expressible.  For example, it
+  should be possible to target Android by something like `$ bazel
+  build --platform=android_armv7`.
+- Configuration settings must be easily expressible either by command
+  line or via a config file (`.bazelrc`, `--bazelrc=file`).
+
+#### Integrated testing
+
+Bazel has builtin support for test targets.
+
+#### Remote caching and execution
+
+Bazel supports both of these, but they are outside the scope of this
+proposal.  Adding Bazel support would enable them.
+
+#### Continuous Integration support
+
+I have no experience with CI systems or processes, but Bazel claims to
+  have good support; e.g. see [Using Bazel in a continuous integration
+  system](https://blog.bazel.build/2016/01/27/continuous-integration.html)
+  and [Remote
+  cache](https://github.com/angular/angular/blob/master/docs/BAZEL.md#remote-cache).
+
+#### Maintainability and Extensibility
+
+It must be easy to maintain the build sytem.
+
+It should be easy to extend the system - for example, to add another
+target platform, toolchain, etc.  In particular, it should be easy to
+introduce new build tools.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+Platform support:
+  * Linux - required, I presume
+  * MacOS - my development environment
+  * Windows - this is a policy decision.  I can design for
+    portability, but I do not have a Windows machine so I cannot test
+    Windows support.  I do have a copy of Parallels, so if somebody
+    can provide me with a licensed copy of Windows I can support it.
+
+### Tasks
+
+General:
+* Minimal build (local monobuild).
+  * WORKSPACE file, BUILD.bazel files containing target specifications, .bzl files containing macros, global vars, custom rules, etc.
+  * `$ bazel build //pkg:target` completes successfully for each target
+  * MacOS first, then Linux, then Windows.
+* Adaptive builds (local monobuilds with platform-specific adaptations).
+  * A Bazel `platform()` is defined for each officially supported platform
+    * Platform definition includes architecture, OS, toolchain.
+  * Build specifications are adapted to each supported platform
+    * For example, the compilers for different platforms take
+      different options.  Bazel does a good job of automatically
+      setting the compile command for each platform, but in some cases
+      additional options should or must be specified.
+  * Bazel does not have `./configure` functionality; it cannot detect
```

Comment:
You can use repository rules with `configure = true` as a proxy for this. Repository rules have unlimited access to analyze the system and generate BUILD files containing `platform` / `toolchain` declarations dynamically.

Check out bazel's own C++ autoconfigure repo rule: https://github.com/bazelbuild/bazel/blob/dcd19067212bfe2fb7eb2c404c2aa0cf3850752b/tools/cpp/cc_configure.bzl#L79-L86

Reply:
Thanks.  I came across https://github.com/bazelbuild/bazel/issues/8786 a few days ago but have not yet looked into it in detail.